File: WinPaletter/GUI/Elements/Input/TextBox.vb
Imports System.ComponentModel
Imports System.Drawing.Drawing2D
Imports System.Drawing.Text

Namespace UI.WP

    <Description("Themed TextBox for WinPaletter UI")> <DefaultEvent("TextChanged")> Public Class TextBox : Inherits Control

        Sub New()
            SetStyle(ControlStyles.AllPaintingInWmPaint Or ControlStyles.UserPaint Or ControlStyles.ResizeRedraw Or ControlStyles.SupportsTransparentBackColor, True)
            DoubleBuffered = True

            ForeColor = Color.White

            TB = New Windows.Forms.TextBox With {
                .Font = New Font("Segoe UI", 9),
                .Text = Text,
                .ForeColor = Color.White,
                .MaxLength = _MaxLength,
                .Multiline = _Multiline,
                .ReadOnly = _ReadOnly,
                .UseSystemPasswordChar = _UseSystemPasswordChar,
                .BorderStyle = BorderStyle.None,
                .Location = New Point(1, 0),
                .Width = Width - 3,
                .Cursor = Cursors.IBeam,
                .ScrollBars = Scrollbars,
                .WordWrap = WordWrap
            }

            If _Multiline Then
                TB.Height = Height - 8
            Else
                Height = TB.Height + 8
            End If

            If My.Style.DarkMode Then BackColor = Color.FromArgb(55, 55, 55) Else BackColor = Color.FromArgb(225, 225, 225)
            If My.Style.DarkMode Then TB.BackColor = Color.FromArgb(55, 55, 55) Else TB.BackColor = Color.FromArgb(225, 225, 225)
        End Sub

#Region "Variables"

        Private WithEvents TB As Windows.Forms.TextBox
        Private _Shown As Boolean = False
        Private ActiveTTLColor As Color

        Private State As MouseState = MouseState.None

        Enum MouseState As Byte
            None = 0
            Over = 1
            Down = 2
            Block = 3
        End Enum

#End Region

#Region "Properties"

        Public Property DrawOnGlass As Boolean = False

        Private _TextAlign As HorizontalAlignment = HorizontalAlignment.Left

        <Category("Options")>
        Property TextAlign() As HorizontalAlignment
            Get
                Return _TextAlign
            End Get
            Set(value As HorizontalAlignment)
                _TextAlign = value
                If TB IsNot Nothing Then
                    TB.TextAlign = value
                End If
            End Set
        End Property

        Private _MaxLength As Integer = 32767

        <Category("Options")>
        Property MaxLength() As Integer
            Get
                Return _MaxLength
            End Get
            Set(value As Integer)
                _MaxLength = value
                If TB IsNot Nothing Then
                    TB.MaxLength = value
                End If
            End Set
        End Property

        Private _ReadOnly As Boolean

        <Category("Options")>
        Property [ReadOnly]() As Boolean
            Get
                Return _ReadOnly
            End Get
            Set(value As Boolean)
                _ReadOnly = value
                If TB IsNot Nothing Then
                    TB.ReadOnly = value
                End If
            End Set
        End Property

        Private _UseSystemPasswordChar As Boolean

        <Category("Options")>
        Property UseSystemPasswordChar() As Boolean
            Get
                Return _UseSystemPasswordChar
            End Get
            Set(value As Boolean)
                _UseSystemPasswordChar = value
                If TB IsNot Nothing Then
                    TB.UseSystemPasswordChar = value
                End If
            End Set
        End Property

        Private _Multiline As Boolean

        <Category("Options")>
        Property Multiline() As Boolean
            Get
                Return _Multiline
            End Get
            Set(value As Boolean)
                _Multiline = value
                If TB IsNot Nothing Then
                    TB.Multiline = value

                    If value Then
                        TB.Height = Height - 8
                    Else
                        Height = TB.Height + 8
                    End If

                End If
            End Set
        End Property

        <Browsable(True)>
        <DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)>
        <EditorBrowsable(EditorBrowsableState.Always)>
        <Editor(GetType(System.ComponentModel.Design.MultilineStringEditor), GetType(System.Drawing.Design.UITypeEditor))>
        <Bindable(True)>
        <Category("Options")>
        Overrides Property Text As String
            Get
                Return MyBase.Text
            End Get
            Set(value As String)
                MyBase.Text = value
                If TB IsNot Nothing Then
                    TB.Text = value
                End If
            End Set
        End Property

        <Category("Options")>
        Overrides Property Font As Font
            Get
                Return MyBase.Font
            End Get
            Set(value As Font)
                MyBase.Font = value
                If TB IsNot Nothing Then
                    TB.Font = value
                    TB.Location = New Point(3, 4)
                    TB.Width = Width - 6

                    If Not _Multiline Then
                        Height = TB.Height + 8
                    End If
                End If
            End Set
        End Property

        Private _Scrollbars As Windows.Forms.ScrollBars = ScrollBars.None
        Public Property Scrollbars As Windows.Forms.ScrollBars
            Get
                Return _Scrollbars
            End Get
            Set(value As Windows.Forms.ScrollBars)
                _Scrollbars = value
                TB.ScrollBars = value
            End Set
        End Property

        Private _WordWrap As Boolean = True
        Public Property WordWrap As Boolean
            Get
                Return _WordWrap
            End Get
            Set(value As Boolean)
                _WordWrap = value
                TB.WordWrap = value
            End Set
        End Property

        Public Property SelectionStart As Integer
            Get
                Return TB.SelectionStart
            End Get
            Set(value As Integer)
                TB.SelectionStart = CInt(value)
            End Set
        End Property

        Public Property SelectionLength As Integer
            Get
                Return TB.SelectionLength
            End Get
            Set(value As Integer)
                TB.SelectionLength = CInt(value)
            End Set
        End Property

        Public Property SelectedText As String
            Get
                Return TB.SelectedText
            End Get
            Set(value As String)
                TB.SelectedText = CStr(value)
            End Set
        End Property

        Protected Overrides ReadOnly Property CreateParams() As CreateParams
            Get
                Dim cpar As CreateParams = MyBase.CreateParams
                If DrawOnGlass And Not DesignMode Then
                    cpar.ExStyle = cpar.ExStyle Or &H20
                    Return cpar
                Else
                    Return cpar
                End If
            End Get
        End Property

#End Region

#Region "Events"

        Public Event KeyboardPress(s As Object, e As KeyPressEventArgs)

        Protected Overrides Sub OnCreateControl()
            MyBase.OnCreateControl()
            If Not Controls.Contains(TB) Then
                Controls.Add(TB)
            End If
        End Sub

        Private Sub OnBaseTextChanged(s As Object, e As EventArgs)
            Text = TB.Text
        End Sub

        Private Sub OnBaseKeyDown(s As Object, e As KeyEventArgs)
            If e.Control AndAlso e.KeyCode = Keys.A Then
                TB.SelectAll()
                e.SuppressKeyPress = True
            End If
            If e.Control AndAlso e.KeyCode = Keys.C Then
                TB.Copy()
                e.SuppressKeyPress = True
            End If
        End Sub

        Public Overloads Sub OnKeyPress(s As Object, e As KeyPressEventArgs)
            RaiseEvent KeyboardPress(s, e)
        End Sub

        Protected Overrides Sub OnResize(e As EventArgs)
            TB.Location = New Point(4, 4)
            TB.Width = Width - 14

            If _Multiline Then
                TB.Height = Height - 8
            Else
                Height = TB.Height + 8
            End If

            MyBase.OnResize(e)
        End Sub

        Protected Overrides Sub OnMouseDown(e As MouseEventArgs)
            MyBase.OnMouseDown(e)
            State = MouseState.Down
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Protected Overrides Sub OnMouseUp(e As MouseEventArgs)
            MyBase.OnMouseUp(e)
            State = MouseState.Over
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            TB.Focus() : Invalidate()
        End Sub

        Protected Overrides Sub OnMouseEnter(e As EventArgs)
            MyBase.OnMouseEnter(e)
            State = MouseState.Over
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Protected Overrides Sub OnMouseLeave(e As EventArgs)
            MyBase.OnMouseLeave(e)
            State = MouseState.None
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Private Sub TB_MouseDown(sender As Object, e As MouseEventArgs) Handles TB.MouseDown
            State = MouseState.Down
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Private Sub TB_MouseEnter(sender As Object, e As EventArgs) Handles TB.MouseEnter, TB.MouseUp
            State = MouseState.Over
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Private Sub TB_MouseLeave(sender As Object, e As EventArgs) Handles TB.MouseLeave
            State = MouseState.None
            _Shown = True
            Timer.Enabled = True
            Timer.Start()
            Invalidate()
        End Sub

        Private Sub TB_LostFocus(sender As Object, e As EventArgs) Handles TB.LostFocus
            State = MouseState.None
            Invalidate()
        End Sub

        Protected Overrides Sub OnHandleCreated(e As EventArgs)
            Try
                If Not DesignMode Then
                    MyBase.OnHandleCreated(e)
                    alpha = 0
                End If
            Catch
            End Try
        End Sub

        Private Sub TextBox_HandleCreated(sender As Object, e As EventArgs) Handles Me.HandleCreated
            alpha = 0
            If Not DesignMode Then
                Try
                    AddHandler FindForm.Load, AddressOf Loaded
                    AddHandler FindForm.Shown, AddressOf Showed
                    AddHandler TB.TextChanged, AddressOf OnBaseTextChanged
                    AddHandler TB.KeyDown, AddressOf OnBaseKeyDown
                    AddHandler TB.KeyPress, AddressOf OnKeyPress
                Catch
                End Try
            End If
        End Sub

        Private Sub TextBox_HandleDestroyed(sender As Object, e As EventArgs) Handles Me.HandleDestroyed
            If Not DesignMode Then
                Try
                    RemoveHandler FindForm.Load, AddressOf Loaded
                    RemoveHandler FindForm.Shown, AddressOf Showed
                    RemoveHandler TB.TextChanged, AddressOf OnBaseTextChanged
                    RemoveHandler TB.KeyDown, AddressOf OnBaseKeyDown
                    RemoveHandler TB.KeyPress, AddressOf OnKeyPress
                Catch
                End Try
            End If
        End Sub

        Sub Loaded(sender As Object, e As EventArgs)
            _Shown = False
        End Sub

        Sub Showed(sender As Object, e As EventArgs)
            _Shown = True
        End Sub

#End Region

#Region "Animator"

        Dim alpha As Integer
        ReadOnly Factor As Integer = 20
        Dim WithEvents Timer As New Timer With {.Enabled = False, .Interval = 1}

        Private Sub Timer_Tick(sender As Object, e As EventArgs) Handles Timer.Tick
            If Not DesignMode Then

                If State = MouseState.Over Then
                    If alpha + Factor <= 255 Then
                        alpha += Factor
                    ElseIf alpha + Factor > 255 Then
                        alpha = 255
                        Timer.Enabled = False
                        Timer.Stop()
                    End If

                    If _Shown Then
                        Threading.Thread.Sleep(1)
                        Invalidate()
                    End If
                End If

                If Not State = MouseState.Over Then
                    If alpha - Factor >= 0 Then
                        alpha -= Factor
                    ElseIf alpha - Factor < 0 Then
                        alpha = 0
                        Timer.Enabled = False
                        Timer.Stop()
                    End If

                    If _Shown Then
                        Threading.Thread.Sleep(1)
                        Invalidate()
                    End If
                End If
            End If
        End Sub

#End Region

        Protected Overrides Sub OnPaint(e As PaintEventArgs)
            Dim G As Graphics = e.Graphics
            DoubleBuffered = True
            G.SmoothingMode = SmoothingMode.AntiAlias
            G.TextRenderingHint = TextRenderingHint.SystemDefault

            MyBase.OnPaint(e)

            Try
                ActiveTTLColor = New VisualStyles.VisualStyleRenderer(VisualStyles.VisualStyleElement.Window.Caption.Active).GetColor(VisualStyles.ColorProperty.TextColor).Invert
            Catch
                ActiveTTLColor = SystemColors.ActiveCaptionText
            End Try

            If Not DrawOnGlass Then
                If My.Style.DarkMode Then
                    If ForeColor <> Color.White Then ForeColor = Color.White
                Else
                    If ForeColor <> Color.Black Then ForeColor = Color.Black
                End If
            Else
                ForeColor = ActiveTTLColor
            End If

            TB.ForeColor = ForeColor

            Dim OuterRect As New Rectangle(0, 0, Width - 1, Height - 1)
            Dim InnerRect As New Rectangle(1, 1, Width - 3, Height - 3)

            Dim ParentColor As Color = MyBase.GetParentColor
            Dim LineNone, LineHovered As Color
            Dim BackNone, BackHovered As Color

            If Not DrawOnGlass Then
                LineNone = If(My.Style.DarkMode, ParentColor.Light(0.3), ParentColor.Light(0.05))
                LineHovered = My.Style.Colors.Border_Checked_Hover

                BackNone = If(My.Style.DarkMode, ParentColor.Light(0.05), ParentColor.Light(0.3))
                BackHovered = My.Style.Colors.Back_Checked
            Else
                LineNone = If(Not ActiveTTLColor.IsDark, ParentColor.Light(0.3), ParentColor.Light(0.05))
                LineHovered = My.Style.Colors.Border_Checked_Hover

                BackNone = If(Not ActiveTTLColor.IsDark, ParentColor.Light(0.05), ParentColor.Light(0.3))
                BackHovered = My.Style.Colors.Back_Checked
            End If

            Dim FadeInColor As Color = Color.FromArgb(alpha, LineHovered)
            Dim FadeOutColor As Color = Color.FromArgb(255 - alpha, LineNone)

            If DrawOnGlass Then
                G.Clear(Color.Transparent)
            Else
                G.Clear(ParentColor)
            End If

            If TB.Focused Or Focused Then
                If Not DrawOnGlass Then
                    Using br As New SolidBrush(BackHovered) : G.FillRoundedRect(br, OuterRect) : End Using
                    Using P As New Pen(LineHovered) : G.DrawRoundedRect_LikeW11(P, OuterRect) : End Using
                    TB.BackColor = BackHovered
                Else
                    Using P As New Pen(LineHovered) : G.DrawRoundedRect_LikeW11(P, OuterRect) : End Using
                    TB.BackColor = ParentColor
                End If

            Else
                If Not DrawOnGlass Then
                    Using br As New SolidBrush(BackNone) : G.FillRoundedRect(br, InnerRect) : End Using
                    Using br As New SolidBrush(Color.FromArgb(alpha, BackNone)) : G.FillRoundedRect(br, OuterRect) : End Using
                    Using P As New Pen(FadeInColor) : G.DrawRoundedRect_LikeW11(P, OuterRect) : End Using
                    Using P As New Pen(FadeOutColor) : G.DrawRoundedRect_LikeW11(P, InnerRect) : End Using
                    TB.BackColor = BackNone
                Else
                    Using P As New Pen(FadeInColor.CB(0.1)) : G.DrawRoundedRect_LikeW11(P, OuterRect) : End Using
                    Using P As New Pen(FadeOutColor.CB(0.1)) : G.DrawRoundedRect_LikeW11(P, InnerRect) : End Using
                    TB.BackColor = ParentColor
                End If
            End If

        End Sub

    End Class

End Namespace

File: WinPaletter/GUI/Elements/Input/TextBox.cs
using FluentTransitions;
using System;
using System.ComponentModel;
using System.ComponentModel.Design;
using System.Drawing;
using System.Drawing.Design;
using System.Drawing.Drawing2D;
using System.Drawing.Text;
using System.Runtime.CompilerServices;
using System.Windows.Forms;

namespace WinPaletter.UI.WP
{
    [Description("Themed TextBox for WinPaletter UI")]
    [DefaultEvent("TextChanged")]
    public class TextBox : Control
    {
        public TextBox()
        {
            TB = new()
            {
                Font = new("Segoe UI", 9f),
                Text = Text,
                ForeColor = ForeColor,
                MaxLength = _MaxLength,
                Multiline = _Multiline,
                ReadOnly = _ReadOnly,
                UseSystemPasswordChar = _UseSystemPasswordChar,
                BorderStyle = BorderStyle.None,
                Cursor = Cursors.IBeam,
                ScrollBars = _Scrollbars,
                WordWrap = _WordWrap,
            };

            // Initialize fixed TB height
            _fixedTextBoxHeight = CalculateTextBoxHeight();

            alpha = 0;

            SetTBSizes();

            SetStyle(ControlStyles.UserPaint | ControlStyles.SupportsTransparentBackColor, true);
            UpdateStyles();

            BackColor = Color.Transparent;
            DoubleBuffered = true;
        }

        #region Variables

        // Fixed height for the inner TextBox (calculated from font)
        private int _fixedTextBoxHeight;
        private int leftOffset => Multiline ? 2 : 6;

        private bool CanAnimate => !DesignMode && Program.Style.Animations && this != null && Visible && Parent != null && Parent.Visible && FindForm() != null && FindForm().Visible;

        public System.Windows.Forms.TextBox TB
        {
            [MethodImpl(MethodImplOptions.Synchronized)]
            get => tb;

            [MethodImpl(MethodImplOptions.Synchronized)]
            set
            {
                if (tb != null)
                {
                    tb.MouseDown -= TB_MouseDown;
                    tb.MouseEnter -= TB_MouseEnter;
                    tb.MouseUp -= TB_MouseUp;
                    tb.MouseLeave -= TB_MouseLeave;
                    tb.LostFocus -= TB_LostFocus;
                    tb.TextChanged -= OnBaseTextChanged;
                    tb.KeyDown -= OnBaseKeyDown;
                    tb.KeyPress -= OnKeyPress;
                }

                tb = value;

                if (tb != null)
                {
                    tb.MouseDown += TB_MouseDown;
                    tb.MouseEnter += TB_MouseEnter;
                    tb.MouseUp += TB_MouseUp;
                    tb.MouseLeave += TB_MouseLeave;
                    tb.LostFocus += TB_LostFocus;
                    tb.TextChanged += OnBaseTextChanged;
                    tb.KeyDown += OnBaseKeyDown;
                    tb.KeyPress += OnKeyPress;
                }
            }
        }
        private System.Windows.Forms.TextBox tb;


        #endregion

        #region Properties

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public override Color BackColor { get; set; }

        private HorizontalAlignment _TextAlign = HorizontalAlignment.Left;

        [Category("Options")]
        public HorizontalAlignment TextAlign
        {
            get => _TextAlign;
            set
            {
                if (value != _TextAlign)
                {
                    _TextAlign = value;
                    if (tb is not null) { tb.TextAlign = value; }
                }
            }
        }

        private int _MaxLength = 32767;

        [Category("Options")]
        public int MaxLength
        {
            get => _MaxLength;
            set
            {
                if (value != _MaxLength)
                {
                    _MaxLength = value;
                    if (tb is not null) { tb.MaxLength = value; }
                }
            }
        }

        private bool _ReadOnly;

        [Category("Options")]
        public bool ReadOnly
        {
            get => _ReadOnly;
            set
            {
                if (value != _ReadOnly)
                {
                    _ReadOnly = value;
                    if (tb is not null) { tb.ReadOnly = value; }
                }
            }
        }

        private bool _UseSystemPasswordChar;

        [Category("Options")]
        public bool UseSystemPasswordChar
        {
            get => _UseSystemPasswordChar;
            set
            {
                if (value != _UseSystemPasswordChar)
                {
                    _UseSystemPasswordChar = value;
                    if (tb is not null) { tb.UseSystemPasswordChar = value; }
                }
            }
        }

        private bool _Multiline;

        [Category("Options")]
        public bool Multiline
        {
            get => _Multiline;
            set
            {
                if (value != _Multiline)
                {
                    _Multiline = value;
                    if (TB is not null)
                    {
                        tb.Multiline = value;

                        // Update TB sizes and anchors based on multiline setting
                        UpdateTextBoxConfiguration();
                    }
                }
            }
        }

        [Browsable(true)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Visible)]
        [EditorBrowsable(EditorBrowsableState.Always)]
        [Editor(typeof(MultilineStringEditor), typeof(UITypeEditor))]
        [Bindable(true)]
        [Category("Options")]
        public override string Text
        {
            get => base.Text;
            set
            {
                if (value != base.Text)
                {
                    base.Text = value;
                    if (TB is not null) { tb.Text = value; }
                }
            }
        }

        [Category("Options")]
        public override Font Font
        {
            get => base.Font;
            set
            {
                if (value != base.Font)
                {
                    base.Font = value;
                    if (tb is not null)
                    {
                        tb.Font = value;

                        // Recalculate fixed height for inner TextBox
                        _fixedTextBoxHeight = CalculateTextBoxHeight();

                        // Update TB configuration
                        UpdateTextBoxConfiguration();
                    }
                }
            }
        }

        private ScrollBars _Scrollbars = ScrollBars.None;
        public ScrollBars Scrollbars
        {
            get => _Scrollbars;
            set
            {
                if (value != _Scrollbars)
                {
                    _Scrollbars = value;
                    tb.ScrollBars = value;
                }
            }
        }

        private bool _WordWrap = true;
        public bool WordWrap
        {
            get => _WordWrap;
            set
            {
                if (value != _WordWrap)
                {
                    _WordWrap = value;
                    tb.WordWrap = value;
                }
            }
        }

        public int SelectionStart
        {
            get => tb.SelectionStart;
            set
            {
                if (value != tb.SelectionStart) { tb.SelectionStart = value; }
            }
        }

        public int SelectionLength
        {
            get => tb.SelectionLength;
            set
            {
                if (value != tb.SelectionLength) { tb.SelectionLength = value; }
            }
        }

        public string SelectedText
        {
            get => tb.SelectedText;
            set
            {
                if (value != tb.SelectedText) { tb.SelectedText = value; }
            }
        }

        #endregion

        #region Events/Overrides

        /// <summary>
        /// Calculates the proper height for the inner TextBox based on current font
        /// </summary>
        private int CalculateTextBoxHeight()
        {
            // For single-line TextBox, calculate height based on font
            // Using "Ay" to get proper height including ascenders and descenders
            return TextRenderer.MeasureText("Ay", TB?.Font ?? this.Font).Height + (_Multiline ? 0 : 3);
        }

        /// <summary>
        /// Updates the TextBox configuration based on current properties
        /// </summary>
        private void UpdateTextBoxConfiguration()
        {
            if (TB == null) return;

            if (_Multiline)
            {
                // For multiline: TB fills the entire container
                TB.Anchor = AnchorStyles.Top | AnchorStyles.Bottom | AnchorStyles.Left | AnchorStyles.Right;
                TB.Location = new Point(leftOffset, 4);
                TB.Height = Math.Max(Height - 8, _fixedTextBoxHeight); // Minimum height is font height
                TB.Width = Width - (leftOffset * 2); // Account for both left and right
            }
            else
            {
                // For single line: fixed height, anchored left/right, centered vertically
                TB.Anchor = AnchorStyles.Left | AnchorStyles.Right;
                TB.Height = _fixedTextBoxHeight;
                TB.Width = Width - (leftOffset * 2); // Account for both left and right

                // Center vertically
                CenterTextBoxVertically(leftOffset);
            }
        }

        /// <summary>
        /// Centers the TextBox vertically within the container
        /// </summary>
        private void CenterTextBoxVertically(int leftOffset)
        {
            if (TB == null || _Multiline) return;

            int y = (Height - TB.Height) / 2;
            // Ensure minimum padding
            y = Math.Max(4, y);
            TB.Location = new Point(leftOffset, y);
        }

        /// <summary>
        /// Updates TB sizes and position
        /// </summary>
        void SetTBSizes()
        {
            if (TB == null) return;

            // Update fixed height
            _fixedTextBoxHeight = CalculateTextBoxHeight();

            // Apply configuration
            UpdateTextBoxConfiguration();
        }

        public event KeyboardPressEventHandler KeyboardPress;

        public delegate void KeyboardPressEventHandler(object s, KeyPressEventArgs e);

        protected override void OnCreateControl()
        {
            if (!Controls.Contains(TB))
            {
                Controls.Add(TB);
                SetTBSizes();
            }

            base.OnCreateControl();
        }

        private void OnBaseTextChanged(object s, EventArgs e)
        {
            Text = tb.Text;
        }

        private void OnBaseKeyDown(object s, KeyEventArgs e)
        {
            if (e.Control && e.KeyCode == Keys.A)
            {
                tb.SelectAll();
                e.SuppressKeyPress = true;
            }
            if (e.Control && e.KeyCode == Keys.C)
            {
                tb.Copy();
                e.SuppressKeyPress = true;
            }
        }

        public void OnKeyPress(object s, KeyPressEventArgs e)
        {
            KeyboardPress?.Invoke(s, e);
        }

        protected override void OnResize(EventArgs e)
        {
            base.OnResize(e);

            // Update TB when container is resized
            if (TB != null)
            {
                if (_Multiline)
                {
                    // Multiline: adjust TB height to fill container
                    TB.Height = Math.Max(Height - 8, _fixedTextBoxHeight);
                    TB.Width = Width - (leftOffset * 2);
                    TB.Location = new Point(leftOffset, 4);
                }
                else
                {
                    // Single line: maintain fixed height, adjust width, and center
                    TB.Height = _fixedTextBoxHeight;
                    TB.Width = Width - (leftOffset * 2);
                    CenterTextBoxVertically(leftOffset);
                }
            }

            Invalidate();
        }

        protected override void OnMouseDown(MouseEventArgs e)
        {
            if (CanAnimate) { Transition.With(this, nameof(alpha), 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 0; }

            base.OnMouseDown(e);
        }

        protected override void OnMouseUp(MouseEventArgs e)
        {
            if (CanAnimate) { Transition.With(this, nameof(alpha), ContainsFocus ? 255 : 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = ContainsFocus ? 255 : 0; }

            tb.Focus();

            base.OnMouseUp(e);
        }

        protected override void OnMouseEnter(EventArgs e)
        {
            if (CanAnimate) { Transition.With(this, nameof(alpha), 255).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 255; }

            base.OnMouseEnter(e);
        }

        protected override void OnMouseLeave(EventArgs e)
        {
            if (CanAnimate) { Transition.With(this, nameof(alpha), 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 0; }

            base.OnMouseLeave(e);
        }

        private void TB_MouseDown(object sender, MouseEventArgs e)
        {
            OnMouseDown(e);
        }

        private void TB_MouseEnter(object sender, EventArgs e)
        {
            OnMouseEnter(e);
        }

        private void TB_MouseUp(object sender, MouseEventArgs e)
        {
            OnMouseUp(e);
        }

        private void TB_MouseLeave(object sender, EventArgs e)
        {
            OnMouseLeave(e);
        }

        private void TB_LostFocus(object sender, EventArgs e)
        {
            if (CanAnimate) { Transition.With(this, nameof(alpha), 0).CriticalDamp(TimeSpan.FromMilliseconds(Program.AnimationDuration)); }
            else { alpha = 0; }
        }

        protected override void OnForeColorChanged(EventArgs e)
        {
            tb.ForeColor = ForeColor;

            base.OnForeColorChanged(e);
        }

        protected override void Dispose(bool disposing)
        {
            base.Dispose(disposing);

            tb?.Dispose();
        }

        int parentLevel = 0;
        protected override void OnParentChanged(EventArgs e)
        {
            base.OnParentChanged(e);

            parentLevel = this.Level();
        }

        #endregion

        #region Animator

        private int _alpha = 0;
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden), Browsable(false)]
        public int alpha
        {
            get => _alpha;
            set { _alpha = value; Invalidate(); }
        }

        #endregion

        protected override void OnPaintBackground(PaintEventArgs pevent)
        {
            //Leave it empty to make control background transparent
            base.OnPaintBackground(pevent);
        }

        protected override void OnPaint(PaintEventArgs e)
        {
            Graphics G = e.Graphics;

            //Makes background drawn properly, and transparent
            InvokePaintBackground(this, e);

            Config.Scheme scheme = Enabled ? Program.Style.Schemes.Main : Program.Style.Schemes.Disabled;

            G.SmoothingMode = SmoothingMode.AntiAlias;
            G.TextRenderingHint = DesignMode ? TextRenderingHint.ClearTypeGridFit : Program.Style.TextRenderingHint;

            Rectangle OuterRect = new(0, 0, Width - 1, Height - 1);
            Rectangle InnerRect = new(1, 1, Width - 3, Height - 3);

            Color Line = scheme.Colors.Line_Checked;
            Color LineHover = scheme.Colors.Back_Hover(parentLevel);
            Color FadeInColor = Color.FromArgb(alpha, Line);
            Color FadeOutColor = Color.FromArgb(255 - alpha, LineHover);

            if (tb.Focused | Focused)
            {
                G.FillRoundedRect(scheme.Brushes.Back_Checked, OuterRect);

                if (Program.Style.DarkMode)
                {
                    G.DrawRoundedRectBeveledReverse(scheme.Pens.Line_Checked_Hover, OuterRect);
                }
                else
                {
                    G.DrawRoundedRectBeveled(scheme.Pens.Line_Checked_Hover, OuterRect);
                }

                tb.BackColor = scheme.Colors.Back_Checked;
            }

            else
            {
                using (SolidBrush br = new(scheme.Colors.Back(parentLevel))) { G.FillRoundedRect(br, InnerRect); }

                using (SolidBrush br = new(Color.FromArgb(alpha, scheme.Colors.Back(parentLevel)))) { G.FillRoundedRect(br, OuterRect); }

                if (Program.Style.DarkMode)
                {
                    using (Pen P = new(FadeInColor)) { G.DrawRoundedRectBeveledReverse(P, OuterRect); }
                    using (Pen P = new(FadeOutColor)) { G.DrawRoundedRectBeveledReverse(P, InnerRect); }
                }
                else
                {
                    using (Pen P = new(FadeInColor)) { G.DrawRoundedRectBeveled(P, OuterRect); }
                    using (Pen P = new(FadeOutColor)) { G.DrawRoundedRectBeveled(P, InnerRect); }
                }

                tb.BackColor = scheme.Colors.Back(parentLevel);
            }
        }
    }
}